components/workspace-rollout-job/pkg/wsbridge/wsbridge.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package wsbridge
+
+import (
+	"context"
+	"fmt"
+
+	logrus "github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/gpctl/pkg/util"
+	"github.com/gitpod-io/gitpod/ws-manager-bridge/api"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	log = logrus.WithField("component", "ws-manager-bridge-client")
+)
+
+// RolloutAction is the interface that wraps the updateScore method.
+type RolloutAction interface {
+	UpdateScore(ctx context.Context, clusterName string, score int32) error
+	GetScore(ctx context.Context, clusterName string) (int32, error)
+}
+
+type WsManagerBridgeClient struct {
+	connection *grpc.ClientConn
+}
+
+func NewWsManagerBridgeClient(ctx context.Context, kubeConfig *rest.Config, localPort int) (*WsManagerBridgeClient, error) {
+	clientSet, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// 8080 is the port of the `ws-manager-bridge` grpc service
+	port := fmt.Sprintf("%d:%d", localPort, 8080)
+	podName, err := util.FindAnyPodForComponent(clientSet, "default", "ws-manager-bridge")
+	if err != nil {
+		return nil, err
+	}
+	readychan, errchan := util.ForwardPort(ctx, kubeConfig, "default", podName, port)
```

Comment:
When doing port-forward manually with kubectl, sometimes I see that it just stops working all of a sudden and only works again if I stop and port-forward again. Is that scenario possible here too?

What would happen to the job if the port-forward suddenly stops working?

Reply:
Good Question!

> Is that scenario possible here too?

From my experiments even both shorter and longer, I have not experienced this with the job.

> What would happen to the job if the port-forward suddenly stops working?

The `UpdateScore` function would return an error, and thus causing the job to crash with a failed status (at that point which the rollout DRI should take manual action) . Considering that this is a failure with `wsbridge` and not prometheus, We don't have much else to do as these errors could be anything and not just port-forward-ing. For these kind of errors with `prometheus`, We handle them much more gracefully by reverting the rollout. 🤔